File: src/MixNetFx/Facesso.GenericControls/CustomListViewGroups.vb
Imports ActiveDev
Imports System.Collections.ObjectModel

Public Class CustomListViewGroup(Of InfoItemType As IInfoItem)
    Private _InfoItems As InfoItems(of InfoItemType)
    Private _GroupName As String

    Sub New(ByVal Groupname As String, ByVal InfoItems As InfoItems(Of InfoItemType))
        _InfoItems = InfoItems
        _GroupName = Groupname
    End Sub

    Public Property InfoItems() As InfoItems(Of InfoItemType)
        Get
            Return _InfoItems
        End Get
        Set(ByVal value As InfoItems(Of InfoItemType))
            _InfoItems = value
        End Set
    End Property

    Public Property GroupName() As String
        Get
            Return _GroupName
        End Get
        Set(ByVal value As String)
            _GroupName = value
        End Set
    End Property

End Class

Public Class CustomListViewGroups(Of InfoItemType As IInfoItem)
    Inherits KeyedCollection(Of String, CustomListViewGroup(Of InfoItemType))

    Private myKeyProvidedThroughAdd As String = Nothing

    Protected Overrides Function GetKeyForItem(ByVal item As CustomListViewGroup(Of InfoItemType)) As String
        Return item.GroupName
    End Function

    Public Function GroupSortIndexOfID(ByVal ID As Integer) As Integer
        Dim locCount As Integer = Me.Count
        For Each item As CustomListViewGroup(Of InfoItemType) In Me
            If item.InfoItems.Contains(New IntKey(ID)) Then
                Return Me.Count - locCount + 1
            End If
            locCount -= 1
        Next
        Return 0
    End Function

End Class

File: src/NetFxCSharp/Facesso.GenericControls/CustomListViewGroups.cs
using ActiveDev;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.GenericControls
{
    public class CustomListViewGroup<InfoItemType>
        where InfoItemType : IInfoItem
    {
        private InfoItems<InfoItemType> _InfoItems;
        private string _GroupName;
        public CustomListViewGroup(string Groupname, InfoItems<InfoItemType> InfoItems)
        {
            _InfoItems = InfoItems;
            _GroupName = Groupname;
        }

        public InfoItems<InfoItemType> InfoItems
        {
            get
            {
                return _InfoItems;
            }

            set
            {
                _InfoItems = value;
            }
        }

        public string GroupName
        {
            get
            {
                return _GroupName;
            }

            set
            {
                _GroupName = value;
            }
        }
    }

    public class CustomListViewGroups<InfoItemType> : KeyedCollection<string, CustomListViewGroup<InfoItemType>> where InfoItemType : IInfoItem
    {
        private string myKeyProvidedThroughAdd = null;
        protected override string GetKeyForItem(CustomListViewGroup<InfoItemType> item)
        {
            return item.GroupName;
        }

        public int GroupSortIndexOfID(int ID)
        {
            int locCount = this.Count;
            foreach (CustomListViewGroup<InfoItemType> item in this)
            {
                if (item.InfoItems.Contains(new IntKey(ID)))
                {
                    return this.Count - locCount + 1;
                }

                locCount -= 1;
            }

            return 0;
        }
    }
}